### DISCUSSION TITLE
MariaDB Query not returning all rows.

### DISCUSSION THREAD

**Discussion by msapiro** [sqlalchemy/sqlalchemy#8955] [Usage Questions] [answered]
SQLAlchemy 1.4.44
MariaDB 10.5.16 and 10.5.18
PyMySQL 0.10.1 and 1.0.2

With the above software versions and model defined at https://gitlab.com/mailman/mailman/-/blob/master/src/mailman/model/mailinglist.py#L101 and the ListManager name_components method defined as
```
    @property
    @dbconnection
    def name_components(self, store):
        """See `IListManager`."""
        result_set = store.query(MailingList)
        for mail_host, list_name in result_set.values(MailingList.mail_host,
                                                      MailingList.list_name):
            yield list_name, mail_host
```
(Full module [here](https://gitlab.com/mailman/mailman/-/blob/master/src/mailman/model/listmanager.py#L135))
and `mailinglist` tables containing more than 10 rows, this code
```
        list_manager = getUtility(IListManager)
        utility = getUtility(IMailTransportAgentAliases)
        by_domain = {}
        sort_key = attrgetter('list_name')
        for list_name, mail_host in list_manager.name_components:
            mlist = _FakeList(list_name, mail_host)
            by_domain.setdefault(mlist.mail_host, []).append(mlist)
```
(Full module [here](https://gitlab.com/mailman/mailman/-/blob/master/src/mailman/mta/postfix.py#L143))
returns entries only for the first 10 rows and throws this warning
```
/usr/lib/python3.9/site-packages/pymysql/connections.py:799: UserWarning: Previous unbuffered result was left incomplete
```
This does not occur with other DBMs nor with MariaDB and SQLAlchemy 1.3.24.

The related Mailman Issue is [here](https://gitlab.com/mailman/mailman/-/issues/1044).

Is this a SQLAlchemy bug?

Note that we can work around this issue by changing
```
        for list_name, mail_host in list_manager.name_components:
```
to
```
        for list_name, mail_host in list(list_manager.name_components):
```

**Comment by zzzeek** [comment, marked as answer]
> ```
>     @property
>     @dbconnection
>     def name_components(self, store):
>         """See `IListManager`."""
>         result_set = store.query(MailingList)
>         for mail_host, list_name in result_set.values(MailingList.mail_host,
>                                                       MailingList.list_name):
>             yield list_name, mail_host
> ```
> 
> 
> (Full module [here](https://gitlab.com/mailman/mailman/-/blob/master/src/mailman/mta/postfix.py#L143)) returns entries only for the first 10 rows and throws this warning
> 
> ```
> /usr/lib/python3.9/site-packages/pymysql/connections.py:799: UserWarning: Previous unbuffered result was left incomplete
> ```

The query.values() method is deprecated in 1.4 so I would recommend mailman move off of it.   What it does is sets the [yield_per](https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.Query.yield_per) option on the Query to ten rows, meaning it will use a server side cursor yielding groups of ten rows at a time.   The warning looks like the loop consuming the iterator is breaking out and not completing its read of the iterator.

> 
> This does not occur with other DBMs nor with MariaDB and SQLAlchemy 1.3.24.

the .values() method works the same way in SQLAlchemy 1.3.24, setting yield_per to 10.

> Is this a SQLAlchemy bug?

unknown without a self contained reproducer, though the workaround you mention below suggests the bug is in mailman.   Overall, moving off of values() will remove the implicit use of a server side cursor.

**Comment by msapiro** [comment]
The actual cause is still a mystery. There is nothing in Mailman that would terminate the `for list_name, mail_host in list_manager.name_components:` loop prematurely. Also [here](https://gitlab.com/mailman/mailman/-/issues/1044#note_1197198509) I suggested this patch
```
--- a/lib/sqlalchemy/dialects/mysql/mariadbconnector.py
+++ b/lib/sqlalchemy/dialects/mysql/mariadbconnector.py
@@ -44,7 +44,7 @@ class MySQLExecutionContext_mariadbconnector(MySQLExecutionContext):
     _lastrowid = None
 
     def create_server_side_cursor(self):
-        return self._dbapi_connection.cursor(buffered=False)
+        return self._dbapi_connection.cursor(buffered=True)
 
     def create_default_cursor(self):
         return self._dbapi_connection.cursor(buffered=True)
```
to create buffered server side cursors, and [here](https://gitlab.com/mailman/mailman/-/issues/1044#note_1197755551) it was reported that didn't help.

In any case, the underlying cause is now moot as I have replaced the deprecated `query.values()` method with `query.with_entities()` and that fixes the issue.

Thank you for your answer. It was helpful.

**Reply by zzzeek** [reply]
> The actual cause is still a mystery. There is nothing in Mailman that would terminate the `for list_name, mail_host in list_manager.name_components:` loop prematurely. Also [here](https://gitlab.com/mailman/mailman/-/issues/1044#note_1197198509) I suggested this patch
> 
> ```
> --- a/lib/sqlalchemy/dialects/mysql/mariadbconnector.py
> +++ b/lib/sqlalchemy/dialects/mysql/mariadbconnector.py
> @@ -44,7 +44,7 @@ class MySQLExecutionContext_mariadbconnector(MySQLExecutionContext):
>      _lastrowid = None
>  
>      def create_server_side_cursor(self):
> -        return self._dbapi_connection.cursor(buffered=False)
> +        return self._dbapi_connection.cursor(buffered=True)
>  
>      def create_default_cursor(self):
>          return self._dbapi_connection.cursor(buffered=True)
> ```
> 
> to create buffered server side cursors, and [here](https://gitlab.com/mailman/mailman/-/issues/1044#note_1197755551) it was reported that didn't help.
> 
> In any case, the underlying cause is now moot as I have replaced the deprecated `query.values()` method with `query.with_entities()` and that fixes the issue.


your stack trace shows pymysql being used, that's a completely different driver.    changing a behavior in the "mariadbconnector" dialect could not possibly have an effect if your database URL uses "mysql+pymysql" or "mariadb+pymysql".

**Reply by msapiro** [reply]
Thanks for the clarification.
